app/models/beta/search/facet_classification.rb

```diff
@@ -0,0 +1,64 @@
+module Beta
+  module Search
+    # Encapsulates a set of facet classifications for a given goods nomenclature
+    #
+    # These enable filtering on search results based on semantic categories (filter[animal_type]=swine).
+    class FacetClassification
+      attr_accessor :classifications
+
+      class << self
+        def build(goods_nomenclature, classifications = {})
+          applicable_facet_classifiers = heading_facet_mappings[goods_nomenclature.heading.short_code]
+
+          goods_nomenclature.classifiable_goods_nomenclatures.each do |gn|
+            tokens_for(gn).each do |token|
+              matching_facet_classifications = word_classifications[token] || {}
+
+              matching_facet_classifications.slice(*applicable_facet_classifiers).each do |facet, classification|
+                classifications[facet] ||= {}
+
+                # We always want the most precise classifications
+                #
+                # In practice this means we bail extending classifications if the lowest goods_nomenclature
+                # node in the tree has found any for the given facet category
+                next if classifications[facet].except(gn.goods_nomenclature_sid).any?
```

Comment:
I guess we could create a private method with a meaningful name for this?

def  classification_for_facet_not_empty?  (I actually don't like much this name 🤔 )
  classifications[facet].except(gn.goods_nomenclature_sid).any?
end